### DISCUSSION TITLE
Problem overriding dependency with a functools.partial(Dep, ...) where Dep takes **kwargs

### DISCUSSION THREAD

**Discussion by estan** [fastapi/fastapi#8817] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import functools

from fastapi import FastAPI, Depends
from fastapi.testclient import TestClient

app = FastAPI()

class Dep:
    def __init__(self, _, **kwargs):
        pass


def get_dep():
    return Dep(1)


@app.get("/")
async def root(dep: Dep = Depends(get_dep)):
    print("root called")
    return "a"


def test_failing():
    app.dependency_overrides[get_dep] = functools.partial(Dep, 1)
    assert TestClient(app).get("/").json() == "a"
    app.dependency_overrides = {}


def test_passing():
    app.dependency_overrides[get_dep] = lambda: Dep(1)
    assert TestClient(app).get("/").json() == "a"
    app.dependency_overrides = {}
```

### Description

I'm attempting to override a dependency with a `functools.partial(Dep, ...)`.

- In the above MRE, the `test_failing` test case is failing with

      E       AssertionError: assert {'detail': [{'loc': ['query', 'kwargs'], 'msg': 'field required', 'type': 'value_error.missing'}]} == 'a'

  and as seen in the output if running with `pytest -s`, the `root` endpoint is never called when the test case runs.
- The `test_passing` test case, which uses a lambda instead of a partial, passes.
- I expect both test cases to pass.

It seems to be due to `Dep` taking `**kwargs`.

### Operating System

Linux

### Operating System Details

Ubuntu 20.04.3 LTS

### FastAPI Version

0.68.0

### Python Version

Python 3.8.10

**Comment by Kludex** [comment]
Can you change the order of those tests?

**Comment by estan** [comment]
Yes, the result is the same.

But now I got a little further and can cut Databases out of the equation with the following revised MRE:

```python
import functools

from fastapi import FastAPI, Depends
from fastapi.testclient import TestClient

app = FastAPI()

class Dep:
    def __init__(self, _, **kwargs):
        pass


def get_dep():
    return Dep(1)


@app.get("/")
async def root(dep: Dep = Depends(get_dep)):
    print("root called")
    return "a"


def test_failing():
    app.dependency_overrides[get_dep] = functools.partial(Dep, 1)
    assert TestClient(app).get("/").json() == "a"
    app.dependency_overrides = {}


def test_passing():
    app.dependency_overrides[get_dep] = lambda: Dep(1)
    assert TestClient(app).get("/").json() == "a"
    app.dependency_overrides = {}
```

This fails in the same way. The problem seems to appear because Dep takes a `**kwargs`.

I'll update issue title and description. EDIT: Done.

**Comment by YuriiMotov** [comment, marked as answer]
When you override dependency by function with `**kwargs`, FastAPI will try to resolve dependencies, find `kwargs` parameter and treat it as a required `Query` parameter.
You don't pass it, so you have error 422.

Just use `lambda: Dep(1)` or create new function that will declare parameters that are really needed and pass them to the `__init__` of `Dep`
